odoo/addons/base/models/ir_mail_server.py

```diff
@@ -579,11 +593,13 @@ def _prepare_email_message(self, message, smtp_session):
             del message['To']           # avoid multiple To: headers!
             message['To'] = x_forge_to
 
-        # Try to not spoof the mail from headers
+        # Try to not spoof the mail from headers; fetch session-based or contextualized
+        # values for encapsulation computation
         from_filter = getattr(smtp_session, 'from_filter', False)
         smtp_from = getattr(smtp_session, 'smtp_from', False) or smtp_from
-
-        notifications_email = email_normalize(self._get_default_from_address())
+        notifications_email = email_normalize(
```

Comment:
to not forget:

the patch to not loose the context in `mock_smtplib_connection`

```py
IrMailServer = self.env['ir.mail_server']

connect_origin = type(IrMailServer).connect
find_mail_server_origin = type(IrMailServer)._find_mail_server

def mock_function(func):
    called = 0

    def _call(*args, **kwargs):
        nonlocal called
        called += 1
        return func(*args, **kwargs)

    _call.assert_called_once = lambda: self.assertEqual(called, 1)
    return _call


with patch('smtplib.SMTP_SSL', side_effect=lambda *args, **kwargs: self.testing_smtp_session), \
     patch('smtplib.SMTP', side_effect=lambda *args, **kwargs: self.testing_smtp_session), \
     patch.object(type(IrMailServer), '_is_test_mode', lambda self: False), \
     patch.object(type(IrMailServer), 'connect', mock_function(connect_origin)) as connect_mocked, \
     patch.object(type(IrMailServer), '_find_mail_server', mock_function(find_mail_server_origin)) as find_mail_server_mocked:
    self.connect_mocked = connect_mocked
    self.find_mail_server_mocked = find_mail_server_mocked
    yield
```

to add in `test_mail_server_send_email`
```py
# custom notification / bounce email from context
context_server = self.env['ir.mail_server'].create({'name': 'context', 'from_filter': 'context.example.com', 'smtp_host': 'test'})
IrMailServer = self.env["ir.mail_server"].with_context(
    notifications_email="notification@context.example.com",
    bounce_address="bounce@context.example.com",
)
with self.mock_smtplib_connection()):
    mail_server, smtp_from = IrMailServer._find_mail_server(email_from='"Name" <test@unknown_domain.com>')
    self.assertEqual(mail_server, context_server)
    self.assertEqual(smtp_from, "notification@context.example.com")
    smtp_session = IrMailServer.connect(smtp_from=smtp_from)  # add mail_server_id=mail_server.id if find_mail_server keep not using the context
    message = self._build_email(mail_from='"Name" <test@unknown_domain.com>')
    IrMailServer.send_email(message, smtp_session=smtp_session)

self.assertEqual(len(self.emails), 1)
self.assertSMTPEmailsSent(
    smtp_from="bounce@context.example.com",
    message_from='"Name" <notification@context.example.com>',
    from_filter=context_server.from_filter,
)
```